csrc/scheduler/transpose.cpp

```diff
@@ -1219,17 +1219,24 @@ void scheduleTranspose(Fusion* fusion, TransposeParams params) {
   reference1->reorder({{inner_most_pos2_in_ref1 + 1, -1}});
   // [..., I1/tile1, .., I2/tile2, ..., tile1, tile2]
 
-  // Merge remaining dimensions
-  int64_t lhs_i = -1;
-  for (int64_t i = reference1->nDims() - 2; i > 0; i--) {
-    auto axis_i = i - 1;
-    if (lhs_i == -1) {
-      lhs_i = axis_i;
-    } else {
-      reference1->merge(axis_i, lhs_i);
-      lhs_i = axis_i;
+  // Merge remaining dimensions ignoring reduction axes (See Issue #2317)
+  // The reduction axes cannot be at any position.
+  // For example: [i0, r1, i1, r2, i2] after tiling is [i0, r1, i1/tile1, r2,
+  // i2/tile2, tile1, tile2] The following code merges all the outer iterdomains
+  // as: [i0 * i1/tile1 * i2/tile2, r1, r2, tile1, tile2]
```

Comment:
one quick question, I think the scheduler assumed to have merged non-reduction domain on the outer most?

```
  reference1->split(0, 1);
  // [merged_dim, 1, tile1, tile2]
```

But if there is a reduction id on the outer most, we are screwed here right?

Reply:
You are right. A cleaner fix would be to move the reduction dimensions before this using `moveReductionsOut` for both references. Then we have established a protocol on where the reduction dimensions will be.
That will also simplify this logic since non-reduction domains are inner to all reductions.

What do you think?

Another thing I am trying to verify is if the same thing happens for broadcasr dimension.

Reply:
```
reference1->axis(1)->parallelize(ParallelType::Unswitch);
reference1->axis(0)->parallelize(ParallelType::BIDx);
```

This will need to be modified to find the non-tile dimensions which will be after the reduction domains.

I think this approach might need more careful changes but would be less error-prone.

Reply:
> A cleaner fix would be to move the reduction dimensions before this using moveReductionsOut for both references. Then we have established a protocol on where the reduction dimensions will be.
That will also simplify this logic since non-reduction domains are inner to all reductions.
What do you think?

except you still need to change that later in `Step 4: Schedule group1`. since now inner most are not the tile ids.

I don't have a strong preference here, we can either do a re-order, or just change the parallelize call below to use the right id (non-reduction).

> Another thing I am trying to verify is if the same thing happens for broadcasr dimension.

good catch!

Reply:
> > A cleaner fix would be to move the reduction dimensions before this using moveReductionsOut for both references. Then we have established a protocol on where the reduction dimensions will be.
> > That will also simplify this logic since non-reduction domains are inner to all reductions.
> > What do you think?
> 
> except you still need to change that later in `Step 4: Schedule group1`. since now inner most are not the tile ids.
> 
> I don't have a strong preference here, we can either do a re-order, or just change the parallelize call below to use the right id (non-reduction).


The outermost will be reduction domains. The inner most should still be tile. My suggestion was basically to do `moveReductionsOut` here instead of later Step 3 and Step 4.

Reply:
> The outermost will be reduction domains. The inner most should still be tile.

sorry I was mistaken. Yeah that sounds good to me.